src/sentry/static/sentry/app/utils/tokenizeSearch.tsx

```diff
@@ -1,93 +1,176 @@
-import flatMap from 'lodash/flatMap';
-
 import {escapeDoubleQuotes} from 'app/utils';
 
-export type QueryResults = {
-  /**
-   * The text portion of the string query
-   */
-  query: string[];
-  [k: string]: string[];
+export enum TokenType {
+  OP,
+  TAG,
+  QUERY,
+}
+
+export type Token = {
+  type: TokenType;
+  key?: string;
+  value: string;
 };
 
-/**
- * Tokenize a search into an object
- *
- * Example:
- *   tokenizeSearch('is:resolved foo bar tag:value');
- *   {
- *     is ['resolved'],
- *     query: ['foo', 'bar'],
- *     tag: ['value'],
- *   }
- *
- * Should stay in sync with src.sentry.search.utils:tokenize_query
- */
-export function tokenizeSearch(query: string) {
-  const tokens = splitSearchIntoTokens(query);
+export class QueryResults {
+  tagValues: object;
+  tokens: Token[];
+
+  constructor(strTokens: string[]) {
+    this.tokens = [];
+    this.tagValues = {};
+    for (let token of strTokens) {
+      let tokenState = TokenType.QUERY;
 
-  const searchParams: {query: string[]; tags: string[]} = {
-    query: [],
-    tags: [],
-  };
+      if (['OR', 'AND'].includes(token.toUpperCase())) {
+        this.addOp(token.toUpperCase());
+        continue;
+      }
+
+      if (token.startsWith('(')) {
+        const parenMatch = token.match(/^\(+/g);
```

Comment:
Do you need to regex if you already know that the first character is a paren based on the check on line 30?

Reply:
Yes because it could match more than one paren.

Reply:
Right. I didn't think of that.